### PR TITLE
Add cast to ((align) - 1) for Alignment macros

### LINKED ISSUE #18253: [glib] Add cast to ((align) - 1) for Alignment macros
### Current Behavior

On 64bit machine, when passing in align as an unsigned int (size is 32bit), for calculation ~(align - 1), bitwise NOT was done to only the 32bit. Later, when calculating the & operation, the left-hand side was a gssize (signed pointer-sized int, which is 64bit). Thus, the result of bitwise NOT operation needed to be cast to 64bit. Since it was an unsigned int type, the higher 32bit was all zeros after the cast. That is incorrect.

### Expected Behavior

The bitwise NOT operation needs to be done on a variable which is the same size as gssize.

## On which platforms did you notice this

[x ] macOS
[x ] Linux
[ ] Windows

**Version Used**:

mono master

### PR DESCRIPTION
**Pull request by fanyang-mono**
Fixes #18253

### CODE REVIEW

**Comment by jaykrell** [comment]
Consider `((type)align)`
instead of `(type)(align - 1)`

or even `(type)((type)align) - 1)`

for clarity but it'll tend not to matter.
i.e. to get align typed prior to the subtraction.

Really these should be template functions, not macros.

**Comment by CoffeeFlux** [comment]
Another somewhat tangential note: when fixing a specific issue, I recommend making a branch in your fork off the latest upstream master and applying the fix there. You avoid having unrelated work in your commit history, this PR being a good example since the only actual diff is the final commit and the rest is noise. 

Alternatively, if you want to just work out of your fork's master branch, I'd reset the HEAD to the commit at the upstream's master instead of the repeated merge commits and rebase on top of that if you have local changes you want to preserve. Helps keep the history clean :)

**Comment by fanyang-mono** [comment]
@monojenkins build failed

**Comment by fanyang-mono** [comment]
Alignment value must be a power of two. 0 is an invalid value to begin with. Maybe we should check the validity of Alignment when using this macro. Additionally, I am okay moving the cast to align directly to make the intention of the cast clear.

**Comment by jaykrell** [comment]
Alignment must be a power of two: But that it is not enforced or asserted or static_asserted anything.
I contend this should be template functions, and go ahead and use % and such, and let the compiler optimize, i.e. for the powers of two. Unless there is a desire to optimize -O0.
Also you can cast both align and 1.
One of the other points is that any expression involving different types, such as int and unsigned, is unclear. Though mixing unsigned types with unsigned types, and signed types with signed types, is reasonable. The type of a literal is not clear though. There are various rules, based on the magnitude of the literal.

**Comment by fanyang-mono** [comment]
As far as I know, C doesn't have template function feature. Macro is the closest alternative.

**Comment by fanyang-mono** [comment]
For llvm function allocateDataSection, when alignment is zero, a default alignment of 16 will be used. I wonder for other usage of these macros, how should we handle the case of alignment equals to zero.

**Comment by CoffeeFlux** [comment]
I really think you're overthinking how much this macro needs to do. If you move the cast and add some tests, I'll happily approve.

**Comment by fanyang-mono** [comment]
> I really think you're overthinking how much this macro needs to do. If you move the cast and add some tests, I'll happily approve.

I was basically trying to resolve the 0 case you brought up earlier. If you consider that as overthinking. Then, I don't think there is a need for moving the cast or add a test case for zero.

BTW, I do plan to add tests for other cases.

File: mono/eglib/test/memory.c [hunk 1]
```diff
@@ -30,8 +30,100 @@ test_memory_zero_size_allocations (void)
...
+static RESULT
+test_align_unsigned_int (void)
+{
+	gssize orig_value = 67;
+	gssize result, exp_value;
+	gpointer orig_ptr = (gpointer)orig_value;
+        gpointer result_ptr, exp_ptr;
+
+	// test case #1
+        unsigned int align = 1;
```
Comment (jaykrell) [inline mono/eglib/test/memory.c]: guint?

File: mono/eglib/test/memory.c [hunk 2]
```diff
@@ -30,8 +30,112 @@ test_memory_zero_size_allocations (void)
...
+test_align_signed_int (void)
+{
+	gssize orig_value = 67;
+	gssize result, exp_value;
+	gpointer orig_ptr = (gpointer)orig_value;
+	gpointer result_ptr, exp_ptr;
+
+	// test case #1
+        int align = 1;
+	result = ALIGN_TO(orig_value, align);
```
Comment (CoffeeFlux) [inline mono/eglib/test/memory.c]: Nit: these should probably be `ALIGN_TO (orig_value, align);`

The indentation in this file also seems to be really wonky, but it was that way beforehand and I can sort it out in a separate PR.

File: mono/eglib/test/memory.c (hunk 1 above)
Comment (fanyang-mono) [reply mono/eglib/test/memory.c]: Are g*** types the recommended ones over default C types?
I see people use both of them...
BTW, do we have these best practices documented anywhere?

File: mono/eglib/test/memory.c (hunk 2 above)
Comment (fanyang-mono) [reply mono/eglib/test/memory.c]: Will change the function call format. 

And will try to fix the indentation problem. No guarantee though, cause it doesn't look aligned in my editor. I wonder if there is an editor which has the same indentation mechanism as GitHub? It is really painful to deal with this.

Comment (CoffeeFlux) [reply mono/eglib/test/memory.c]: The issue is just that it mixes spaces and tabs; converting everything to tabs should solve the problem for the given file. It's hard to get an editor that's going to universally find these issues because whether or not they appear aligned is dependent on how many spaces are used for indentation compared with how many your editor is using per-tab.

Github uses 8-space tabs, and my editor uses 4, so the issue with memory.c was very obvious in my editor. On other files though, it has 4 spaces in the file, and so it looks fine in my editor and broken on Github... :(

File: mono/eglib/test/memory.c (hunk 1 above)
Comment (CoffeeFlux) [reply mono/eglib/test/memory.c]: In general I think the policy is to avoid it for things like `char`, but for numerical types use the `guint32` form since it's shorter and unambiguous? @lambdageek can probably comment, but no, I don't think it's properly documented anywhere.

File: mono/eglib/test/memory.c (hunk 2 above)
Comment (fanyang-mono) [reply mono/eglib/test/memory.c]: nvm, git diff actually shows indentation problem

Comment (CoffeeFlux) [reply mono/eglib/test/memory.c]: It looks like the file is still inconsistent overall, though your diff itself is mostly fixed. There are a ton of trailing spaces after the later  `<>ALIGN_TO`, `<> ALIGN_DOWN_TO`, and `<> ALIGN_PTR_TO` lines for some reason, and lines 109/114 have a trailing tab.

We should probably run a linter for this sort of thing on PR diffs, at least for trailing whitespace - I believe CoreCLR/CoreFX has one already? So maybe in the consolidated repo? I know I've accidentally committed trailing whitespace before and discovered it after the fact :\

File: mono/eglib/test/memory.c (hunk 1 above)
Comment (lambdageek) [reply mono/eglib/test/memory.c]: TL;DR - if you know you definitely want a 32- or 64-bit number, use the C99 `int32_t`/`uint32_t`/`int64_t`/`uint64_t` types.  If you just want an `int` use an `int`.

The `gint32`/etc types are from before we had good C99 support from all the compilers we cared about.

I'm not totally convinced we need the `g` types for things where good types exist in the standard now.  The one exception seems to be `gboolean` which we're using pervasively and we're not using `bool` or `_Bool` from C99 anywhere - so in that one case we seem to prefer `gboolean`.

You may also come across `MonoBoolean` - that's used in structures and function signatures of functions that are called from C# - the C# `bool` type corresponds to `MonoBoolean`, not `gboolean`.

Comment (fanyang-mono) [reply mono/eglib/test/memory.c]: gotcha

File: mono/eglib/test/memory.c (hunk 2 above)
Comment (fanyang-mono) [reply mono/eglib/test/memory.c]: The format looks pretty good to me now.

File: mono/eglib/test/memory.c (hunk 1 above)
Comment (jaykrell) [reply mono/eglib/test/memory.c]: `guint` is shorthand for `unsigned`; or consider `guint32`
`gushort` is shorthand for `unsigned short`; or consider `guint16`
`guchar` is shorthand for `unsigned char`; or consider `guint8`

The following are pointless:
 `gchar` just use `char`
 `gint` just use `int`, or consider `gint32`, but really `int` suffices
 `gshort` just use `short`, or consider `gint16`
 `glong` use `long` or really avoid altogether, it is too variably defined across platforms

`gint32` questionable, could be `int`; 16bit `int` is irrelevant
`gint8` could be `int8_t` but it is longer; etc.
`MonoBoolean` would be `bool` if using C++.

Comment (jaykrell) [reply mono/eglib/test/memory.c]: Another wrinkle is `g` is not to be used in public headers, but `int64_t` can be.
I'd suggest really cleanup the code base mechanically and thoroughly and drive out the `g` types.

**Comment by jaykrell** [comment]
The LLVM handling of 0 should not be related to this code.
0 should be invalid here.

**Comment by fanyang-mono** [comment]
@monojenkins build failed
